content/blog/writing-better-angular-tests/index.md

```diff
@@ -0,0 +1,208 @@
+	published: "2020-05-09T04:45:30.247Z",
+	authors: ["skatcat31"],
+	tags: ["testing", "unit tests", "angular"],
+	attached: [],
+	license: "cc-by-nc-sa-4"
+}
+---
+
+There are some evangelicals who say that before code ever exists there always needs to be a test to know how the code should be written. That frankly isn't true. A test isn't *strictly* needed to determine how to code. What **is** needed are tests that give confidence that as code is written a change to already existing functionality doesn't happen and that new functionality will behave properly as time goes on. To this end a lot of testing libraries and frameworks exist. Often times tests are written in regards to the library or framework used and not to the end product's specifications. For Angular this is especially true when the default testing implementation is for testing angular, and not for testing what a developer would use Angular to build. **Tests should be written in the same way a user would use them.** We don't need to test Angular, we need to test what we make with Angular.
+
+# Writing tests for an Angular application does not mean testing Angular
+
+In regards to Angular and writing tests first it must be understood what the tests are for. For a great many projects that means testing a webpage. In proper testing for a webpage the underlying library should be able to be changed at any time for maintainability purposes and the tests should still work. To that end we must write tests for the web and not for Angular. When using the Angular CLI it sets up some tests but when looking closely at the tests it becomes apparent that the tests are testing Angular and not the ouput.
+
+```js
+it('should create the app', () => {
+  const fixture = TestBed.createComponent(AppComponent);
+  const app = fixture.componentInstance;
+  expect(app).toBeTruthy();
+});
+```
+
+This test isn't a very good test. It doesn't say anything about the actual output of the application component itself. When the output is a full rich webpage and tests are testing Angular then the tests won't do much when the content of the webpage is changed. 
+
+While the default testing setup does allow for the writing of tests that would test the outputted HTML they are still specific to Angular
+
+```js
+it('should render title', () => {
+  const fixture = TestBed.createComponent(AppComponent);
+  fixture.detectChanges();
+  const compiled = fixture.nativeElement;
+  expect(compiled.querySelector('.content span').textContent).toContain('The app is running!');
+});
+```
+
+That test looks a little better but it's still very tied to Angular, require in depth knowledge of how Angular actually routes and moves all the bits around to write tests for it and now the tests are completely tied into Angular and the current API footprint. If over the years Angular is retired these tests are now no longer valid.

```

Comment:
```suggestion
That test looks a little better, but it's still very tied to Angular. The test requires in-depth knowledge of how Angular actually routes and moves all the bits around to write tests for it, and as a result, the tests are completely tied into Angular and the current API footprint. If — over the years — Angular is retired, these tests will no longer be valid.
```

I tried not to edit too much here. The first sentence is sorta a run-on and the latter half is easier to read with a bit of grammatical changing